common/browser.go

```diff
@@ -464,6 +464,36 @@ func (b *Browser) NewPage(opts goja.Value) api.Page {
 	return browserCtx.NewPage()
 }
 
+// On returns a Promise that is resolved when the browser process is disconnected.
+// The only accepted event value is "disconnected".
+func (b *Browser) On(event string) *goja.Promise {
+	if event != EventBrowserDisconnected {
+		k6Throw(b.ctx, "unknown browser event: %q, must be %q", event, EventBrowserDisconnected)
+	}
+
+	rt := k6common.GetRuntime(b.ctx) //nolint: staticcheck // SA1019 (deprecated)
+	rcb := GetRegisterCallback(b.ctx)
+	p, resolve, reject := rt.NewPromise()
+	cb := rcb()
+
+	go func() {
+		select {
+		case <-b.browserProc.lostConnection:
+			cb(func() error {
+				resolve(true)
+				return nil
+			})
+		case <-b.ctx.Done():
+			cb(func() error {
+				reject(fmt.Errorf("browser.on promise rejected: %w", b.ctx.Err()))
+				return nil
+			})
```

Comment:
> Another question is about the value passed to reject(). I wasn't sure how to create a JS Error object or whether Goja will properly convert this Go error, and whether we should even return ctx.Err() in this case...

It doesn't need to be error so you can pass w/e. I guess making an error is best and will try to look into it. 

> I haven't been able yet to trigger this failure handler externally..

Does it trigger the other case at all? Do you hit ^C twice?

Reply:
I did add some `fmt.Println` in this go code and a `sleep` before the closing the `browser.close()` and it works as expected:
on success I get: 
```
here
hereing (00m02.2s), 1/1 VUs, 0 complete and 0 interrupted iterations
lost connection

running (00m04.3s), 0/1 VUs, 1 complete and 0 interrupted iterations
default ✓ [======================================] 1 VUs  00m04.3s/10m0s  1/1 iters, 1 per VU

     ✓ should be connected after launch
     ✓ should be connected before ending iteration
     ✓ should be disconnected on event
     ✓ the browser.on success handler should be called

     checks...............: 100.00% ✓ 4        ✗ 0
     data_received........: 0 B     0 B/s
     data_sent............: 0 B     0 B/s
     iteration_duration...: avg=4.28s min=4.28s med=4.28s max=4.28s p(90)=4.28s p(95)=4.28s
     iterations...........: 1       0.233322/s
     vus..................: 1       min=1      max=1
     vus_max..............: 1       min=1      max=1
```

on failure you get: 
```
here
hereing (00m02.2s), 1/1 VUs, 0 complete and 0 interrupted iterations
^Cctx.Done

running (00m02.9s), 0/1 VUs, 0 complete and 1 interrupted iterations
default ✗ [--------------------------------------] 1 VUs  00m02.9s/10m0s  0/1 iters, 1 per VU
WARN[0003] No script iterations finished, consider making the test duration longer

     ✓ should be connected after launch
     ✓ should be connected before ending iteration

     checks..........: 100.00% ✓ 2   ✗ 0
     data_received...: 0 B     0 B/s
     data_sent.......: 0 B     0 B/s
     vus.............: 1       min=1 max=1
     vus_max.........: 1       min=1 max=1

ERRO[0003] err:websocket: close 1006 (abnormal closure): unexpected EOF  category="Connection:handleIOError" elapsed="0 ms" goroutine=98
```

As I explained in the event loop PR - while we wait for all register callbacks to be called that doesn't mean that you will be able to execute code. For now the same context is used for everything and k6 interrupts the goja VM once it's done, so while there is *some* chance for race where you manage to execute the `catch` in this particular case it's not big ;)

This will likely be worked on in future improvements

Reply:
In a addition currently even if you decide to clear the interrupt k6 has *probably* already gotten an error and is only waiting for the register callbacks to callback and will *not* even try to execute them. 

Again this in practice requires that we have separate mechanism to say:
- an iteration is ending (due to an exception for example)
- the scenario is ending 
- the process is ending
- (possibly) gracefulShutdown/Rampdown has been entered. In practice for gracefulRampDown we might actually also want to signal that it's no longer relevant. There is  a case where going up to 10 down to 0 and back up to 10 VUs in cycles might mean that gracefulRampDown doesn't actually stop a VU outright as by the time it has run out we are on the next "up to 10 VUs". 

In all of the above cases we can potentially let javascript code and extensions to do ... something before we completely shutdown the system (as we do now). But we will still need a way to interrupt the goja interpreter as otherwise malicious code can always block k6. Figuring out if goja is executing malicious code or not is also in practice very hard. I am not even certain we can even guess if it's executing code *at all*. 

Arguably a "quick" fix for your case might be to add a bit of a wait on the interrupt :thinking: . Although this will still not work for if I have add `sleep(2)` before the `browser.close()` as then the `sleep` will *not* stop so even if you queue something it will still not get executed ... 

This also kind of illustrates the general problem - we likely want to abort some code in this situation ... but not others. And while most of the code is running synchronously this means the synchronous code blocks any execution of any asynchronous possibly trying to do something. 

So yeah this is an open problem and one that won't be fixed easily. 

Currently waiting on `ctx.Done` is generally useful if you guys want to clean something up without the use of javascript code. For example closing browser or flushing metrics - although that last one usually depends on the context not being close so ... yeah.

Reply:
As always, the wisdom of words, thanks 🙇 

What about this?

```golang
func (b *Browser) On(event string) *goja.Promise {
	...
	go func() {
		a := func() error {
			resolve(true)
			return nil
		}
		r := func() error {
			reject(fmt.Errorf("browser.on promise rejected: %w", b.ctx.Err()))
			return nil
		}
		select {
		case <-b.browserProc.lostConnection:
			select {
			case <-b.ctx.Done():
				cb(r)
				return
			default:
				cb(a)
			}
		case <-b.ctx.Done():
			cb(r)
		}
	}()
```

*RESULT:*

```shell
ERRO[0002] promise rejected: browser.on promise rejected: context canceled  source=console

     ✓ should be connected after launch
     ✓ should be connected before ending iteration
     ✗ the browser.on success handler should be called
      ↳  0% — ✓ 0 / ✗ 1
```

This is most probably not a solution to this problem but it might give us an idea.

Reply:
I noticed something:

`Browser.On`'s `case <-b.ctx.Done():` is selected at the end of the script when we press CTRL+C before the test run ends:

```js
export default function() {
  ...
  sleep(5); // hit CTRL+C here
  browser.close();
}
```

Even though it executes the rejection callback in the `Browser.On`, nothing happens on the k6 output as if it was not rejected at all. Without CTRL+C, the accept callback runs (and context is closed).

This brings us to:

> As I explained in the event loop PR - while we wait for all register callbacks to be called that doesn't mean that you will be able to execute code.

Reply:
@inancgumus We should definitely do that, but from my tests it doesn't help with the interruption. I wasn't able to get a Promise rejection on `^C` even after liberally sprinkling `time.Sleep()` everywhere. Not sure why it works for you... :confused:

Ah, it works if I remove `sleep()` from the script like Mihail mentioned, even without `time.Sleep()` in Go. But that's a general context cancellation, not what happens with `^C`. Though it might be easy to write a test for that case, and if we attach VU to the context and hackily implement the `GetRegisterCallback()` mock, even a unit test should be easy to write. :tada: Will explore this a bit...

But yeah, we need some changes in k6 for the `^C` case to work. @MStoykov A solution might be to pass extensions a cancelable context [here](https://github.com/grafana/k6/blob/fb320f6df8a1694477d96d2f57d982fae1c3d2d1/js/runner.go#L153), and simply cancel it when appropriate. Additionally/alternatively, we could allow extensions to specify a callback on registration (for "deregistration"/cleanup), and run that when appropriate. Or, if we don't want to allow extensions to block k6 (though this shouldn't be a concern; extensions can already do that in many different ways), we could pass them a cancelable context and a "shutdown" channel exposed on `VU` with which they can signal when they're ready to shutdown, and k6 could timeout and do the current forced shutdown after a (user configurable?) amount of time.

Reply:
+1 for this one:

> allow extensions to specify a callback on registration (for "deregistration"/cleanup), and run that when appropriate

Reply:
@imiric please open an issue with your suggestions :)

Reply:
The issue is https://github.com/grafana/k6/issues/2432.

@inancgumus I decided against implementing [this suggestion](https://github.com/grafana/xk6-browser/pull/268#discussion_r822478618), since it fails the current `browser_on.js` test by rejecting the promise (i.e. what we wanted to reproduce with `^C`). I think that if the `lostConnection` case happens, we don't need to do another `ctx.Done()` check to reject the promise. If the connection is closed it should be a successful scenario, and the promise should be rejected only if that's not the case.

Reply:
It's OK. It wasn't a suggestion, but hopefully, a meant to be a way to give you (us) an idea of what was happening :)